js/require fails inside binding

This one is very weird, even though I have it down to a relatively small test case.

Basically, `js/require` fails inside a `let` binding, unless the module was previously loaded at top level.

I reported this previously at https://github.com/bhauman/lein-figwheel/issues/649, when I thought it was Figwheel issue. You may find some value in my earlier ramblings there. But, I think I was faked out by the error being reported by `figwheel-bridge.js`; I now think it is more related to the React Native tooling.

Details:

This project was started with `lein new expo` and I've reduced it down to just a few lines of my code in this branch: https://github.com/deg/receipts-native/tree/figwheel-require.

The key lines are at https://github.com/deg/receipts-native/blob/figwheel-require/src/receipts_native/core.cljs#L11-L26.

The code, as is, crashes. Restoring the commented-out line 15 lets the code work.

````
;;; Restoring this next line removes the crash.
;;; It seems that Figwheel (or something) is unhappy when module is first required in the
;;; expansion of a macro???

;; (defonce native-base (js/require "native-base"))

(defn get-class
       "Extract React class from JavaScript module and adapt as Reagent component."
       [module class-name]
     (r/adapt-react-class (aget module class-name)))

(let [module (js/require "native-base")]
  (def Body      (get-class module "Body"))
  (def Container (get-class module "Container"))
  (def Header    (get-class module "Header"))
  (def Title     (get-class module "Title")))
````

Two errors are reported. The most significant bit seems to be the doubled `js.js` in terminal window error, below.

In the Chrome console, I see:

````
ExceptionsManager.js:73 Error: Script error.
    at syncImportScripts (figwheel-bridge.js:131)
    at importJs (figwheel-bridge.js:148)
    at goog.writeScriptTag_ (figwheel-bridge.js:245)
    at Object.goog.importScript_ (base.js:951)
    at Object.goog.writeScripts_ (base.js:1394)
    at goog.require (base.js:706)
    at figwheel-bridge.js:215
    at syncImportScripts (figwheel-bridge.js:134)
    at importJs (figwheel-bridge.js:148)
    at figwheel-bridge.js:210
reactConsoleErrorHandler @ ExceptionsManager.js:73
console.error @ YellowBox.js:69
syncImportScripts @ figwheel-bridge.js:136
importJs @ figwheel-bridge.js:148
goog.writeScriptTag_ @ figwheel-bridge.js:245
goog.importScript_ @ base.js:951
goog.writeScripts_ @ base.js:1394
goog.require @ base.js:706
(anonymous) @ figwheel-bridge.js:215
syncImportScripts @ figwheel-bridge.js:134
importJs @ figwheel-bridge.js:148
(anonymous) @ figwheel-bridge.js:210
syncImportScripts @ figwheel-bridge.js:134
importJs @ figwheel-bridge.js:148
loadApp @ figwheel-bridge.js:207
componentDidMount @ figwheel-bridge.js:87
chainedFunction @ factory.js:615
commitLifeCycles @ ReactNativeFiber-dev.js:2218
commitAllLifeCycles @ ReactNativeFiber-dev.js:2504
_invokeGuardedCallback @ ReactNativeFiber-dev.js:75
invokeGuardedCallback @ ReactNativeFiber-dev.js:49
commitAllWork @ ReactNativeFiber-dev.js:2530
workLoop @ ReactNativeFiber-dev.js:2596
_invokeGuardedCallback @ ReactNativeFiber-dev.js:75
invokeGuardedCallback @ ReactNativeFiber-dev.js:49
performWork @ ReactNativeFiber-dev.js:2634
scheduleUpdateImpl @ ReactNativeFiber-dev.js:2768
scheduleUpdate @ ReactNativeFiber-dev.js:2751
scheduleTopLevelUpdate @ ReactNativeFiber-dev.js:3048
updateContainer @ ReactNativeFiber-dev.js:3059
render @ ReactNativeFiber-dev.js:4483
renderApplication @ renderApplication.js:33
run @ AppRegistry.js:116
runApplication @ AppRegistry.js:191
__callFunction @ MessageQueue.js:306
(anonymous) @ MessageQueue.js:108
__guard @ MessageQueue.js:269
MessageQueue.callFunctionReturnFlushedQueue @ MessageQueue.js:107
(anonymous) @ debuggerWorker.js:72
10:50:17.515 ExceptionsManager.js:73 Error: Uncaught TypeError: Cannot read property 'call' of undefined
    at syncImportScripts (figwheel-bridge.js:131)
    at importJs (figwheel-bridge.js:148)
    at goog.writeScriptTag_ (figwheel-bridge.js:245)
    at Object.goog.importScript_ (base.js:951)
    at Object.goog.writeScripts_ (base.js:1394)
    at goog.require (base.js:706)
    at figwheel-bridge.js:215
    at syncImportScripts (figwheel-bridge.js:134)
    at importJs (figwheel-bridge.js:148)
    at figwheel-bridge.js:210
````

In the terminal running `exp start --android`, I see:

````
[exp] Launching Dev Tools...
[exp] Dependency graph loaded.
[exp] Finished building JavaScript bundle in 2045ms.
[exp] You are now debugging remotely, check your browser console for your application logs.
[exp] Failed building JavaScript bundle.
[exp] NotFoundError: Cannot find entry file target/expo/goog/base.js.js in any of the roots: ["/home/deg/Documents/git/projects/receipts-native"]
[exp]     at DependencyGraph._getAbsolutePath (/home/deg/Documents/git/projects/receipts-native/node_modules/metro-bundler/src/node-haste/DependencyGraph.js:305:11)
[exp]     at DependencyGraph.getDependencies (/home/deg/Documents/git/projects/receipts-native/node_modules/metro-bundler/src/node-haste/DependencyGraph.js:283:4236)
[exp]     at Resolver.getDependencies (/home/deg/Documents/git/projects/receipts-native/node_modules/metro-bundler/src/Resolver/index.js:129:5)
[exp]     at /home/deg/Documents/git/projects/receipts-native/node_modules/metro-bundler/src/Bundler/index.js:642:39
[exp]     at Generator.next (<anonymous>)
[exp]     at step (/home/deg/Documents/git/projects/receipts-native/node_modules/metro-bundler/src/Bundler/index.js:13:1336)
[exp]     at /home/deg/Documents/git/projects/receipts-native/node_modules/metro-bundler/src/Bundler/index.js:13:1496
[exp]     at <anonymous>
[exp]     at process._tickCallback (internal/process/next_tick.js:188:7)
[exp] NotFoundError: Cannot find entry file target/expo/goog/base.js.js in any of the roots: ["/home/deg/Documents/git/projects/receipts-native"]
[exp]     at DependencyGraph._getAbsolutePath (/home/deg/Documents/git/projects/receipts-native/node_modules/metro-bundler/src/node-haste/DependencyGraph.js:305:11)
[exp]     at DependencyGraph.getDependencies (/home/deg/Documents/git/projects/receipts-native/node_modules/metro-bundler/src/node-haste/DependencyGraph.js:283:4236)
[exp]     at Resolver.getDependencies (/home/deg/Documents/git/projects/receipts-native/node_modules/metro-bundler/src/Resolver/index.js:129:5)
[exp]     at /home/deg/Documents/git/projects/receipts-native/node_modules/metro-bundler/src/Bundler/index.js:642:39
[exp]     at Generator.next (<anonymous>)
[exp]     at step (/home/deg/Documents/git/projects/receipts-native/node_modules/metro-bundler/src/Bundler/index.js:13:1336)
[exp]     at /home/deg/Documents/git/projects/receipts-native/node_modules/metro-bundler/src/Bundler/index.js:13:1496
[exp]     at <anonymous>
[exp]     at process._tickCallback (internal/process/next_tick.js:188:7)
````

--- a/resources/leiningen/new/expo/env/dev/user.clj
+++ b/resources/leiningen/new/expo/env/dev/user.clj
@@ -125,6 +125,15 @@
         (catch Exception e
           (println "Error: " e)))))
 
+(defn- required-modules
+  "returns a vector of string with the names of the imported modules. Ignoring those
+  that are commented out"
+  [file-content]
+  (some->> file-content
+           (re-seq #"(?m)^[^;\n]+?\(js/require \"([^\"]+)\"\)")
+           (map last)
+           (vec)))
+
   ;; Each file maybe corresponds to multiple modules.
   (defn watch-for-external-modules
     []
@@ -144,19 +153,7 @@
 
                                   (when (.exists file)
                                     (let [content (slurp file)
-                                          js-modules (some->>
-                                                       content
-                                                       (re-seq #"\(js/require \"([^\"]+)\"\)")
-                                                       (map last)
-                                                       (vec))
-                                          commented-modules (some->>
-                                                              content
-                                                              (re-seq #"[;]{2,}.*\(js/require \"([^\"]+)\"\)")
-                                                              (map last)
-                                                              (set))
-                                          js-modules (if commented-modules
-                                                       (vec (remove commented-modules js-modules))
-                                                       js-modules)]
+                                          js-modules (required-modules content)]
                                       (let [old-js-modules (get m file-name)]
                                         (when (not= old-js-modules js-modules)
                                           (let [new-m (if (seq js-modules)
@@ -180,19 +177,7 @@
           (let [file-name (-> (.getPath file)
                               (str/replace (str (System/getProperty "user.dir") "/") ""))
                 content (slurp file)
-                js-modules (some->>
-                             content
-                             (re-seq #"\(js/require \"([^\"]+)\"\)")
-                             (map last)
-                             (vec))
-                commented-modules (some->>
-                                    content
-                                    (re-seq #"[;]{2,}.*\(js/require \"([^\"]+)\"\)")
-                                    (map last)
-                                    (set))
-                js-modules (if commented-modules
-                             (vec (remove commented-modules js-modules))
-                             js-modules)]
+                js-modules (required-modules content)]
             (if js-modules
               (swap! m assoc file-name (vec js-modules))))))
       (spit path @m)

